### PR TITLE
Use arviz for plotting

### PR DESCRIPTION
**Pull request by ColCarroll**
This starts moving pymc3 plotting over to arviz. Other things I'd like to include in this PR:

1. Update `setup.py` to include arviz as optional (`pip install pymc3[all]` or `pip install pymc3[plots]`?)
2. Rerun at least a few notebooks to make sure the graphs are _actually_ well supported.

This is still WIP because of https://github.com/arviz-devs/arviz/pull/529

I am also making an issue for `plot_posterior_predictive_glm`, which is heavily used in the GLM notebooks, but not yet supported by arviz. I think there's interest in doing so, though!

Here's an example running under this PR (_and_ the arviz PR). Note that as a side effect of this change, `traceplot` is more flexible.

image

### CODE REVIEW

**Comment by twiecki** [comment]
image

**Comment by eigenfoo** [comment]
Deleted code is debugged code!!

File: pymc3/plots/__init__.py [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+    import arviz as az
+except ImportError:  # arviz is optional, throw exception when used
+
+    class _ImportWarner:
+        def __init__(self, attr):
+            self.attr = attr
+
+        def __call__(self, *args, **kwargs):
+            raise ImportError(
+                f"ArviZ is not installed. In order to use `{self.attr}`:\npip install arviz"
```
Comment (aloctavodia) [inline pymc3/plots/__init__.py]: not sure if we can use fstrings.

**Comment by aloctavodia** [comment]
The documentation will be updated as part of this PR, or a separated one?

**Comment by canyon289** [comment]
Wellll, the code has just shifted :)

But it has become more general purpose which is still a win!

**Comment by ColCarroll** [comment]
- Fixed f-string
- The API docs for plots will now include a link to the ArviZ documentation
- `pip install pymc3[plots]` will include arviz, otherwise it remains optional

File: pymc3/plots/__init__.py [hunk 2]
```diff
@@ -1,9 +1,37 @@
...
+            )
+
+    class _ArviZ:
+        def __getattr__(self, attr):
+            return _ImportWarner(attr)
+
+    az = _ArviZ()
+
+
+autocorrplot = az.plot_autocorr
```
Comment (fonnesbeck) [inline pymc3/plots/__init__.py]: Its tempting to reconcile the names, but I suppose we should just leave things be.

Comment (ColCarroll) [reply pymc3/plots/__init__.py]: from running through the first notebook, `varnames --> var_names` is the worst right now. I am a little tempted to add a decorator just for that case:

```python
def map_args(func):
    @functools.wraps(func)
    def wrapped(*args, **kwargs):
        if 'varnames' in kwargs and 'var_names' not in kwargs:
            warnings.warn('Keyword argument `varnames` renamed to `var_names`, and will be removed in pymc3 3.8.0')
            kwargs['var_names'] = kwargs.pop('varnames')
        return func(*args, **kwargs)
    return wrapped
```

Thoughts on that?

Comment (twiecki) [reply pymc3/plots/__init__.py]: I think that's a good idea.

**Comment by twiecki** [comment]
As the plan is to eventually remove this, we should probably also raise deprecation warnings. Ideally with the new call if that's not too cumbersome to do.

**Comment by ColCarroll** [comment]
Wait... _is_ the plan to eventually remove this? I guess in favor of just using `arviz` directly?

**Comment by fonnesbeck** [comment]
I think we should have `pm.plot_trace` (for example) that is pure Arviz and then have a mapping to `pm.traceplot` that converts the arguments as well, so that it behaves just as `traceplot` currently does. 

I don't think we should deprecate `traceplot`, it should just remain available, but the documentation will describe `plot_trace` going forward. This will allow us to make the changes we want without breaking old code (or littering output with deprecations).

**Comment by ColCarroll** [comment]
ah, neat -- I think I can just dump `az.plots.__all__` into `pm.plot.__all__` to get the "pure arviz" part of things. Then we can do custom stuff to `pm.traceplot` and the like.

**Comment by ColCarroll** [comment]
Sorry for the long pause. I addressed these two issues:

- both original names and arviz names work now. if we want to customize a plot, we can override the pymc3 name
- I put a small deprecation warning function in (the same one from this discussion)

Here's a traceplot using both `pm.plot_trace` and `pm.traceplot(..., var_names=...)`:

image

**Comment by fonnesbeck** [comment]
Is this still a WIP?

**Comment by ColCarroll** [comment]
Oops, changed the title -  I think it is ready. Tab completion and docs come from arviz, and look ok to me.

I will mess with the tests tomorrow to get them passing. I haven't played around too much without arviz installed, which might be causing CI problems.

**Comment by twiecki** [comment]
```
+++pip install -e .
Obtaining file:///home/travis/build/pymc-devs/pymc3
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/travis/build/pymc-devs/pymc3/setup.py", line 66
        tests_require=test_reqs,
                    ^
    SyntaxError: invalid syntax
    
    ----------------------------------------
```

File: setup.py [hunk 3]
```diff
@@ -60,5 +60,8 @@ def get_version():
           include_package_data=True,
           classifiers=classifiers,
           install_requires=install_reqs,
+          extras_require={
+              "plots": ["arviz"],
+          }
```
Comment (twiecki) [inline setup.py]:
```suggestion
          },
```

**Comment by twiecki** [comment]
Looks great. Needs a prominent line in the release notes.

We should probably also fix the docs, but certainly can do that in a separate PR.

**Comment by junpenglao** [comment]
We should have a docathon week for update all docs ;-)

**Comment by ColCarroll** [comment]
Looks like there was a surprising test failure because we are now picking up numpy 1.16.

**Comment by ColCarroll** [comment]
This looks ready to merge again. There was a travis failure from trying to send data to coveralls.

**Comment by ericmjl** [comment]
I'm hitting the big green button!

**Comment by ericmjl** [comment]
Thanks @ColCarroll!

**Comment by twiecki** [comment]
Niceeeee!

**Pull request merged**